src/Symfony/Contracts/Service/ServiceSubscriberTrait.php

```diff
@@ -50,13 +50,15 @@ public static function getSubscribedServices(): array
                 throw new \LogicException(sprintf('Cannot use "%s" on methods without a return type in "%s::%s()".', SubscribedService::class, $method->name, self::class));
             }
 
-            $serviceId = $returnType instanceof \ReflectionNamedType ? $returnType->getName() : (string) $returnType;
+            $attribute = $attribute->newInstance();
 
-            if ($returnType->allowsNull()) {
-                $serviceId = '?'.$serviceId;
-            }
+            /* @var SubscribedService $attribute */
+
+            $attribute->key ??= self::class.'::'.$method->name;
+            $attribute->type ??= $returnType instanceof \ReflectionNamedType ? $returnType->getName() : (string) $returnType;
+            $attribute->nullable = $returnType->allowsNull();
```

Comment:
to maximise BC/FC, I'd suggest not using the attribute unless it's required to convey the reference (aka there are actual attributes)